File: zyx/Leetcode/src/easy/Code35.java
package easy;

/**
 * 35: 搜索插入位置
 * 
 * 给定一个排序数组和一个目标值，在数组中找到目标值，并返回其索引。如果目标值不存在于数组中，返回它将会被按顺序插入的位置。
 * 
 * 请必须使用时间复杂度为 O(log n) 的算法。
 * 
 * 来源：力扣（LeetCode） 链接：https://leetcode-cn.com/problems/search-insert-position
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * 
 * @author tony
 *
 */
public class Code35 {
	public int searchInsert(int[] nums, int target) {
		// 求出数组的长度
		int n = nums.length;
		// 本题从下标0开始，因此最小值为0，最大值为n-1，因为要插入一个数，要记录插入的位置，因此需要一个额外变量来保存这个下标
		int left = 0, right = n - 1, ans = n;
		// 因为从0开始，因此要些<=号
		while (left <= right) {
			// 求中数，防止溢出
			int mid = ((right - left) / 2) + left;
			// <=这个条件，也可以将>=写在一起，但是=与大小于合并在哪里，就要将ans写在哪里，用来记录mid值
			if (target <= nums[mid]) {
				ans = mid;
				right = mid - 1;
			} else {
				left = mid + 1;
			}
		}
		return ans;
	}
}

File: wendy/Leetcode/easy/code35.js
/*
题目：给定一个排序数组和一个目标值，在数组中找到目标值，并返回其索引。如果目标值不存在于数组中，返回它将会被按顺序插入的位置。
     请必须使用时间复杂度为 O(log n) 的算法。
*/
/*
思路：顺序数组-->二分查找法：始末区间，中间值，缩小区间遍历；
     当没有找到的时候，right指针的后一位就是需要插入的位置；
     当插入的数比数组中最后一个数值还大的时候，那么当前的right就是插入点；
     
*/
var searchInsert = function(nums, target) {
    let left = 0, right = nums.length-1
    while(left <= right) { // 判断循环终止条件，什么时候遍历完
        let mid = Math.floor(left + (right - left) / 2)  // 定义中值，防止溢出
        if (target > nums[mid]) { // 将目标值与中值进行判断，缩小区间
            left = mid + 1
        }else if (target < nums[mid]) {
            right = mid -1
        }else {
            return mid
        }
    }
    return right == nums.length ? right : right + 1
};